burp-dradis.rb

```diff
@@ -447,48 +436,68 @@ def create_dradis_issue(issue)
     token    = @field_token.text    || ''
     payload  = build_json_payload(issue)
 
-    path = ''
-    path << @field_path.text || '' if @radio_pro.selected
-    path << '/' unless path[-1,1] == '/'
-    path << 'api/issues'
-
     unless endpoint.length > 0 && token.length > 0
       javax.swing.JOptionPane.showMessageDialog(nil, "Please configure the extension using the #{META::TAB_CAPTION} tab.")
       return
     end
 
     begin
-      uri      = URI.parse(endpoint)
-      http     = Net::HTTP.new(uri.host, uri.port)
-      request  = Net::HTTP::Post.new(path)
-
-      request['Content-Type']      = 'application/json'
+      t = build_http_request_thread(endpoint, token, payload)
+      t.start
+    rescue Exception => e
+      @callbacks.issue_alert("There was an error connecting to Dradis: #{e.message}.")
+      @stderr.println e.backtrace
+    end
+  end
 
+  # Internal: builds a Thread object that makes an HTTP request using the Burp API
+  #
+  # endpoint - The configured Dradis endpoint
+  # token - The configured Dradis API token
+  # payload - The HTTP request body to be sent
+  #
+  # Returns a Thread object
+  #
+  def build_http_request_thread endpoint, token, payload
+    port, use_ssl=
       if endpoint =~ /\Ahttps/i
-        http.use_ssl     = true
-        http.verify_mode = OpenSSL::SSL::VERIFY_NONE if @check_ignore_ssl_errors.selected
+        [443, true]
+      else
+        [80, false]
       end
 
+    path = endpoint.clone
+    path << @field_path.text || '' if @radio_pro.selected
+    path << '/' unless path[-1,1] == '/'
+    path << 'api/issues'
+
+    auth_header =
       if @radio_pro.selected
-        request['Authorization']     = "Token token=\"#{token}\""
-        request['Dradis-Project-Id'] = @field_project_id.text
+        "Authorization: Token token=\"#{token}\"\n" \
+        "Dradis-Project-Id: #{@field_project_id.text}"
       else
-        request.basic_auth('BurpExtender', token)
+        basic_token = Base64.getEncoder().encodeToString("BurpExtender:#{token}".to_java_bytes)
+        "Authorization: Basic #{basic_token}"
       end
 
-      request.body = payload
+    return java.lang.Thread.new(Proc.new {
+      host = endpoint.sub(/^https?\:\/\//, '')
+      url = java.net.URL.new(path)
 
+      # Build the initial request
```

Comment:
I reimagined this part like this:

```ruby
# Build initial request
req_as_bytes  = @helpers.toggleRequestMethod(@helpers.build_http_request(url))
req_as_string = String.from_java_bytes(req_as_bytes)
 
# Edit the request headers
req_as_string.sub! "Content-Length: 0", "Content-Length: #{payload.length}"
req_as_string.sub! "Content-Type: x-www-form-urlencoded", "Content-Type: application/json"
req_as_string.sub! "application/json", "application/json\n#{auth_headers}" #FIXME: this seems unclear to me

# Add body
req_as_string << payload

# Send request
response = @callbacks.make_http_request(host, port, use_ssl, req_as_string.to_java_bytes)
```
I  was trying to:
- avoid overwriting `report_req` variable, I got confused with that
- make this line more clear: `request_str.sub!('x-www-form-urlencoded', "json\n#{auth_header}")` Didin't achieve that though. Ideas welcome. Found this link: https://github.com/codewatchorg/bypasswaf/blob/master/bypasswaf.py